source/Dictionary.cpp

```diff
@@ -44,48 +47,96 @@ namespace {
 		return make_pair(low, false);
 	}
 
-	// String interning: return a pointer to a character string that matches the
-	// given string but has static storage duration.
-	const char *Intern(const char *key)
+	// Creates new if missing.
+	static dict NameToKey(const string &name)
 	{
-		static set<string> interned;
 		static mutex m;
 
-		// Just in case this function is accessed from multiple threads:
 		lock_guard<mutex> lock(m);
-		return interned.insert(key).first->c_str();
+		map<string, dict>::iterator it = nameToKey.lower_bound(name);
+		if(it == nameToKey.end() || nameToKey.key_comp()(name, it->first))
+		{
+			it = nameToKey.insert(it, pair<string, dict>(name, names.size()));
+			names.push_back(it->first.c_str());
+		}
+		return it->second;
 	}
 }
 
 
 
-double &Dictionary::operator[](const char *key)
+double &Dictionary::operator[](const char *name)
+{
+	pair<size_t, bool> pos = Search(name, *this);
+	if(pos.second)
+		return store.data()[data()[pos.first]];
+
+	dict key = NameToKey(name);
+	Grow();
+
+	insert(begin() + pos.first, key);
+	return store[key];
+}
+
+
+
+double &Dictionary::operator[](const string &name)
+{
+	return (*this)[name.c_str()];
+}
+
+
+
+void Dictionary::Update(const char *name, const double &value)
 {
-	pair<size_t, bool> pos = Search(key, *this);
+	pair<size_t, bool> pos = Search(name, *this);
 	if(pos.second)
-		return data()[pos.first].second;
+		store.data()[pos.first] = value;
+}
+
+
+
+void Dictionary::Update(const string &name, const double &value)
+{
+	return Update(name.c_str(), value);
+}
+
 
-	return insert(begin() + pos.first, make_pair(Intern(key), 0.))->second;
+
+double Dictionary::Get(const string &name) const
+{
+	pair<size_t, bool> pos = Search(name.c_str(), *this);
+	return pos.second ? store.data()[data()[pos.first]] : 0.;
+}
+
+
+
+void Dictionary::Grow()
+{
+	if(store.size() <= names.size())
```

Comment:
Why by default grow to the maximum possible size? This will make each Dictionary as large as all possible attribute values. Isn't a lower value sufficient?